Rework FV Lagrange multiplier constraint system

GiudGiud opened the pull request:
adds the point value constraint @grmnptr 
adds a bounding constraint for variables @joe61vette

GiudGiud committed:
Rename FVScalarLagrangeMultiplier to allow for a base class for FV constraints, refs #20607

GiudGiud committed:
Add a point value constraint for FV variables, refs #20607{

Fixup

fixup

GiudGiud committed:
Add a kernel that creates bound on FV variable values, refs #20607

GiudGiud committed:
Add tests for FV constraints, refs #20607

GiudGiud committed:
Add documentation to all FV constraints, refs #20607

GiudGiud commented on framework/doc/content/source/fvkernels/FVScalarLagrangeMultiplier.md:
```diff
@@ -1,22 +0,0 @@
-# FVScalarLagrangeMultiplier
-
-This object implements the residuals that enforce the constraint
```
I swear I used git mv :(

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/990454/) on dc418b9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20610/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20610/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f93b0f3aaf859f125d82ad8d0d513b85dfa4309e`

GiudGiud committed:
More explicit names for bound types for the FVBoundedValueConstraint, refs #20607

Clang format

Update test/tests/fvkernels/constraints/tests

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update framework/src/fvkernels/FVPointValueConstraint.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update framework/src/fvkernels/FVPointValueConstraint.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update framework/doc/content/source/fvkernels/FVBoundedValueConstraint.md

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update framework/src/fvkernels/FVPointValueConstraint.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/990484/) on 4fcccf5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20610/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20610/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f93b0f3aaf859f125d82ad8d0d513b85dfa4309e`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/990513/) on c920057 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20610/site)

This comment will be updated on new commits.

lindsayad commented on framework/include/fvkernels/FVScalarLagrangeMultiplierConstraint.h:
```diff
@@ -0,0 +1,38 @@
+  void computeResidual() override final;
+  void computeJacobian() override final;
+  void computeOffDiagJacobian() override final;
+  ADReal computeQpResidual() override;
```
is this used anymore? I suppose the code coverage report will tell me

lindsayad replied:
I prefer the compile time error that comes with a non-implemented pure virtual (from `FVElementalKernel`)

lindsayad commented on test/tests/fvkernels/constraints/bounded_value.i:
```diff
@@ -0,0 +1,73 @@
+
+[Executioner]
+  type = Transient
+  solve_type = 'PJFNK'
```
why is this PJFNK? And no off-diagonal coupling in the preconditioner? If the LM is playing a role in this problem, I would expect convergence to not be very good without that coupling

lindsayad commented on test/tests/fvkernels/constraints/tests:
```diff
@@ -0,0 +1,30 @@
+    input = 'bounded_value.i'
+    exodiff = 'bounded_value_out.e'
+
+    requirement = 'The system shall be able to impose bounds for the values of of a variable.'
```
```suggestion
    requirement = 'The system shall be able to impose bounds for the values of a variable.'
```

lindsayad commented on framework/doc/content/source/fvkernels/FVBoundedValueConstraint.md:
```diff
@@ -0,0 +1,32 @@
+# FVBoundedValueConstraint
```
This name to me implies boundedness everywhere, so I think maybe something like `FVPointBoundedValueConstraint` is better. Then it seems like we are introducing a canonical naming scheme where classes are either `FVPoint` or `FVIntegral`. (This isn't relevant to your code but relevant to the names here ... we have a `Bounds` system that applies everywhere and that is used in conjunction with PETSc's variational inequality system)

lindsayad replied:
Now that I've gone and looked at the source, I see that this is looking at more than just one point, so I'd adjust the following documentation. This is a very interesting approach that I've never seen before. You are applying a spatially *constant* source/sink such that the primal field never violates the bounds. The approach I am used to is applying a spatially *varying* source/sink (e.g. a *field* Lagrange multiplier) such that the primal field never violates the bounds. Since neither is physical, I suppose I'm fine with this approach I'm not familiar with. @fdkong what do you think about this? Have you ever seen this before?

lindsayad replied:
However, I would guess that you will be introducing, in an integral sense, a much larger source/sink to the system than you would with a spatially varying LM field

lindsayad commented on framework/src/fvkernels/FVPointValueConstraint.C:
```diff
@@ -0,0 +1,67 @@
+
+  // We communicate the results and if there is conflict between processes,
+  // the minimum cell ID is chosen
+  dof_id_type elem_id = elem ? elem->id() : DofObject::invalid_id;
```
```suggestion
  const dof_id_type elem_id = elem ? elem->id() : DofObject::invalid_id;
```

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/991816/) on 560bfd8 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7c8de4</th>
      <th colspan=4>#20610 560bfd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.53%</td>
      <td>82.54%</td>
      <td>+0.01%</td>
      <td>95.12%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73619</td>
      <td>73671</td>
      <td>+52</td>
      <td>78</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15583</td>
      <td>15582</td>
      <td>-1</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20610/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20610/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20610/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
Comments should be addressed

lindsayad commented on framework/doc/content/source/fvkernels/FVBoundedValueConstraint.md:
```diff
@@ -0,0 +1,32 @@
+
+This object implements the residuals that enforce the constraint
+
+!equation
+\phi (point P) > \phi_0
+
+or
+
+!equation
+\phi (point P) < \phi_0
```
I like what you have in the doxygen for this class fine. You could also just reproduce that. Basically I just think we should remove the `(point P)` language

lindsayad commented on framework/src/fvkernels/FVPointValueConstraint.C:
```diff
@@ -0,0 +1,67 @@
+  : FVScalarLagrangeMultiplierConstraint(parameters),
+    _phi0(getParam<Real>("phi0")),
+    _point(getParam<Point>("point")),
+    _my_elem(NULL)
```
```suggestion
    _my_elem(nullptr)
```

lindsayad commented:
I wish that you could mark a conversation as "resolved for this PR" or something 😄 I don't like that you have basically have to collapse all the dialogue boxes. I suppose that's incentive for creating an issue instead which is a good thing

lindsayad commented:
Lol, the CI hadn't run. Ninja merge!

GiudGiud commented:
oh weird it showed "xx checks have passed" and green, not anything running
I ll pay attention next time
